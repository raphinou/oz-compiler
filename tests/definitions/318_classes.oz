%Generic classes
  % from http://www.mozart-oz.org/home/doc/tutorial/node10.html#section.classes.parametrized
  % See also CTMCP 7.4.3
local
   SortClass Int Rat ListC
in
       class ListC from BaseObject
       meth append(Xs Ys $)
          case Xs
          of nil then Ys
          [] X|Xr then
             X|(ListC , append(Xr Ys $))
          end
       end
       meth member(X L $)
          {Member X L}    % This defined in List.oz
       end
       meth length(Xs $)
          case Xs
          of nil then 0
          [] _|Xr then
            (ListC , length(Xr $)) + 1
          end
       end
       meth nrev(Xs ?Ys)
          case Xs
          of nil then Ys = nil
          [] X|Xr then Yr in
             ListC , nrev(Xr Yr)
             ListC , append(Yr [X] Ys)
          end
       end
    end
   fun {SortClass Type}
      class $ from BaseObject
         meth qsort(Xs Ys)
            case Xs
            of nil then Ys = nil
            [] P|Xr then S L in
               {self partition(Xr P S L)}
               ListC, append({self qsort(S $)} P|{self qsort(L $)} Ys)
            end
         end
         meth partition(Xs P Ss Ls)
            case Xs
            of nil then Ss = nil Ls = nil
            [] X|Xr then Sr Lr in
               if Type,less(X P $) then
        Ss=X|Sr
        Lr = Ls
               else
        Ss = Sr
        Ls = X|Lr
               end
               {self partition(Xr P Sr Lr)}
            end
         end
      end
   end

   class Int
      meth less(X Y $)
         X<Y
      end
   end
   class Rat from BaseObject
      meth less(X Y $)
        '/'(P Q) = X
        '/'(R S) = Y
         in
        P*S < Q*R
      end
   end

{Show {{New {SortClass Int} noop} qsort([1 2 5 3 4] $)}}
{Show {{New {SortClass Rat} noop}
     qsort(['/'(23 3) '/'(34 11) '/'(47 17)] $)}}


end
