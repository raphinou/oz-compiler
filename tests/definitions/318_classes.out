[1 2 3 4 5]
['/'(47 17) '/'(34 11) '/'(23 3)]
